### PR TITLE
[FPGA] MemoryBuffering Transformation

### PR DESCRIPTION
**Pull request by Walon1998**
Extends the StreamingMemory transformation to support memory buffering. 
The MemoryBuffering transformation allows reading/writing data from memory
using a wider data format (e.g. 512 bits), and then convert it on the fly to the right data type used by the computation

### CODE REVIEW

File: dace/transformation/dataflow/streaming_memory.py [hunk 1]
```diff
@@ -6,10 +6,12 @@
 import networkx as nx
 import warnings
 
+import dace
```
Comment (TizianoDeMatteis) [inline dace/transformation/dataflow/streaming_memory.py:9]: It is better to import just what you need

File: dace/transformation/dataflow/streaming_memory.py [hunk 2]
```diff
@@ -200,6 +216,86 @@ def can_be_applied(graph: SDFGState,
...
+
+            # Last stride has to be one
+            if strides[-1] != 1:
+                return False
+
+            vector_size = int(self.memory_buffering_target_bytes /
+                              desc.dtype.bytes)
+            strides.pop()  # Remove last element since we already checked it
+
+            # Other strides has to be divisiable by vector size
```
Comment (TizianoDeMatteis) [inline dace/transformation/dataflow/streaming_memory.py:243]: typo

File: dace/transformation/dataflow/streaming_memory.py [hunk 3]
```diff
@@ -263,13 +359,100 @@ def apply(self, sdfg: SDFG) -> nodes.AccessNode:
...
+
+                arrname = str(self.access(sdfg))
+
+                # Add gearbox
+                total_size = edge.data.volume
+                vector_size = int(self.memory_buffering_target_bytes /
+                                  desc.dtype.bytes)
+
+                if isinstance(sdfg.arrays[dnode.data].shape[-1], dace.symbol):
+                    warnings.warn(
```
Comment (TizianoDeMatteis) [inline dace/transformation/dataflow/streaming_memory.py:373]: Maybe we should try to resolve the symbol instead and fail if this is not the case. @definelicht what do you think?

File: tests/fpga/streaming_memory_test.py [hunk 4]
```diff
@@ -237,6 +313,380 @@ def test_streaming_and_composition():
     return sdfg
 
 
+@xilinx_test()
```
Comment (TizianoDeMatteis) [inline tests/fpga/streaming_memory_test.py:316]: What happens if you test this (and the following) also with intel_fpga?

Comment (Walon1998) [reply tests/fpga/streaming_memory_test.py:316]: Timed out: [vec_add, mat_add, tensor_add, multistream, multistream_with_deps, map_order]
Compiler failure: [mapnests]
Pass: [vec_add_1, vec_add_1_symbolic, mat_add_symbol]

**Comment by Walon1998** [comment]
Compiler error if the following vector types are used: float16, int8, int64

**Comment by Walon1998** [comment]
I think that some of the code I added could be replaced by the vectorization transformation once it is complete.

**Pull request closed without merging**